sklearn/linear_model/huber.py

```diff
@@ -0,0 +1,347 @@
+    n_samples, n_features = X.shape
+    fit_intercept = n_features + 2 == w.shape[0]
+    if fit_intercept:
+        intercept = w[-2]
+    sigma = w[-1]
+    w = w[:X.shape[1]]
+
+    if sample_weight is not None:
+        n_samples = np.sum(sample_weight)
+
+    # Calculate the values where |y - X'w -c / sigma| > epsilon
+    # The values above this threshold are outliers.
+    linear_loss = y - safe_sparse_dot(X, w)
+    if fit_intercept:
+        linear_loss -= intercept
+    abs_linear_loss = np.abs(linear_loss)
+    outliers_mask = (abs_linear_loss / sigma) > epsilon
+
+    # Calculate the linear loss due to the outliers.
+    # This is equal to (2 * M * |y - X'w -c / sigma| - M**2)*sigma
+    outliers = abs_linear_loss[outliers_mask]
+    num_outliers = np.count_nonzero(outliers)
+    num_non_outliers = X.shape[0] - num_outliers
+
+    if sample_weight is None:
+        n_outliers = np.count_nonzero(outliers_mask)
+        outlier_loss = (
+            2 * epsilon * np.sum(outliers) -
+            sigma * n_outliers * epsilon**2)
+    else:
+        outliers_sw = sample_weight[outliers_mask]
+        n_outliers = np.sum(outliers_sw)
+        outlier_loss = (
+            2 * epsilon * np.sum(outliers_sw * outliers) -
+            sigma * n_outliers * epsilon**2)
+
+    # Calculate the quadratic loss due to the non-outliers.-
+    # This is equal to |(y - X'w - c)**2 / sigma**2|*sigma
+    non_outliers = linear_loss[~outliers_mask]
+
+    if sample_weight is None:
+        squared_loss = np.dot(non_outliers.T, non_outliers) / sigma
+    else:
+        weighted_non_outliers = sample_weight[~outliers_mask] * non_outliers
+        weighted_loss = np.dot(weighted_non_outliers.T, non_outliers)
+        squared_loss = weighted_loss / sigma
+
+    if fit_intercept:
+        grad = np.zeros(n_features + 2)
+    else:
+        grad = np.zeros(n_features + 1)
+
+
+    # Gradient due to the squared loss.
+    X_non_outliers = -_axis0_safe_slice(X, ~outliers_mask, num_non_outliers)
+    if sample_weight is None:
+        grad[:n_features] = (
+            2 * safe_sparse_dot(non_outliers, X_non_outliers) / sigma)
+    else:
+        grad[:n_features] = (
+            2 * safe_sparse_dot(weighted_non_outliers, X_non_outliers) / sigma)
+
+    # Gradient due to the linear loss.
+    outliers_pos = np.logical_and(linear_loss >= 0, outliers_mask)
+    outliers_neg = np.logical_and(linear_loss < 0, outliers_mask)
+    num_outliers_pos = np.count_nonzero(outliers_pos)
+    num_outliers_neg = num_outliers - num_outliers_pos
+    X_outliers_pos = _axis0_safe_slice(X, outliers_pos, num_outliers_pos)
+    X_outliers_neg = _axis0_safe_slice(X, outliers_neg, num_outliers_neg)
+
+    if sample_weight is None:
+
+        # Summing along any axis of a matrix returns a matrix object.
+        outliers_pos_sum = np.squeeze(np.array(X_outliers_pos.sum(axis=0)))
+        outliers_neg_sum = np.squeeze(np.array(X_outliers_neg.sum(axis=0)))
```

Comment:
That's what I would have thought as well.. But summing up a sparse matrix returns a sparse matrix, rather than an array. So we convert it into a 2-D array and squeeze it into 1-D.

    X_csr = sparse.rand(5, 5, format='csc')
    X_csr.sum(axis=0)
    matrix([[0., 0., 0., 0., 0.]])